### ISSUE TITLE
Memoization of past solutions to speed up future calculations

### ISSUE THREAD

**Issue by dglynch** [teelolws/TLDRMissions#73]
TLDR version: Whenever the addon finds a valid solution to a particular mission, it should memoize the result in the WTF folder, so that, on future days, the calculation of that same mission can check those solution(s) first instead of starting over from scratch.

Sometimes it takes over 300,000 iterations of the calculation algorithm to find a solution to a particular mission or set of missions, and this work needs to be repeated each time the same mission re-appears on the mission table. This can waste several minutes of time every day. Once the addon has solved a particular mission once, it should record and re-use that solution the next time the same mission re-appears. Assuming the same champions and followers are available, the same solution should work again. If for any reason the memoized solution does not work anymore, the addon could fall back to its current behaviour.

See: https://en.wikipedia.org/wiki/Memoization

**Comment by teelolws** [comment]
This isn't that simple. Followers level up and have different stats. They can have different HP at the start of the mission - if they just came from a mission they're often near-death, and the longer the player leaves them to send them back out, the closer to full HP they are. Those both change the outcome of the combatlog.

I currently already do Memoization on a per-session basis, so if a player cancels a long simulation and restarts it, the results will be found much faster. But to move it between sessions, they'll have to be saved using their baseAttack and health, and theres the filesize limit of the saved variables to consider, too.

**Comment by teelolws** [comment]
Okay its coming in the next update. When the results of a mission is cached, it still has to weed through the cache with all the permutations. Its not going to be an instant thing to find a solution, but definitely much faster than having to recreate the combatlog each time.

Try not to take the "Simulations" counter too seriously. I let it run to 20,000 on a difficult mission then reloaded. It skipped through to about 10,000 then continued normally. That doesn't mean its re-simulating 10,000 permutations - it just means the counter never increased for about 10,000 of the permutations that were already cached.

I'm also adding a Slider to let you crank up the simulations to process per frame. It comes with a warning that you're using it at your own risk, so if you set it to 200 processes/frame and your game crashes, don't blame me!
